keanu-project/src/main/java/io/improbable/keanu/network/BayesianNetwork.java

```diff
@@ -18,15 +15,35 @@
 public class BayesianNetwork {
 
     private final List<? extends Vertex> vertices;
+    private final Map<String, Vertex> vertexLabels;
 
     public BayesianNetwork(Set<? extends Vertex> vertices) {
         this.vertices = ImmutableList.copyOf(vertices);
+        this.vertexLabels = getLabels(vertices);
     }
 
     public BayesianNetwork(Collection<? extends Vertex> vertices) {
         this(new HashSet<>(vertices));
     }
 
+    public Vertex getVertexByLabel(String label) {
+        return vertexLabels.get(label);
+    }
+
+    private Map<String, Vertex> getLabels(Set<? extends Vertex>vertices) {
+        Map<String, Vertex> labelMap = new HashMap<>();
+        for (Vertex v : vertices) {
```

Comment:
You can do this more concisely with streams. This doesn't include the check for duplicates but you could do that easily enough by checking the size of the map.

        Map<String, ? extends Vertex> labelMap = vertices.stream()
            .filter(v -> v.getLabel() != null)
            .collect(Collectors.toMap(Vertex::getLabel, v -> v));